docs/src/main/asciidoc/cdi.adoc

```diff
@@ -0,0 +1,361 @@
+* In rare cases it's practical to destroy the beans manually. A direct injected reference would lead to a stale bean instance.
+ 
+ 
+== _Q: OK. You said that there are several kinds of beans?_
+
+A: Yes. In general, in Quarkus we distinguish:
+
+1. Class beans
+2. Producer methods
+3. Producer fields
+4. Synthetic beans
+
+Producer methods are useful if you need additional control over instantiation of a bean.
+They are also useful when integrating third-party libraries where you don't control the class source and may not add additional annotations etc.
+
+.Producer Method Example
+[source,java]
+----
+@ApplicationScoped
+public class PiProducer {
+
+    @Produces <1>
+    double producePi() { 
+       return Math.PI; <2>
+    }
+}
+
+@ApplicationScoped
+public class PiConsumer {
+
+   @Inject
+   double pi; <3>
+   
+   double twoPis() {
+     return pi * 2;
+   }
+} 
+----
+<1> The container analyses the method annotations to build a bean metadata.
+The return type is used to build the set of bean types. 
+In this case, it will be `double` and `java.lang.Object`.
+No scope annotation is declared and so it's defaulted to `@Dependent`.
+<2> The container will delegate to this method when creating the bean instance.
+<3> `PiProducer#producePi()` is invoked and the result is set to `PiConsumer#pi` when the `PiConsumer` instance is created.
+
+NOTE: One of the limitations of producers is that interceptors are not applied to the result of a producer method/field.
+
+Producer fields are very similar to producer methods except that the field value is read when obtaining a bean instance.
+
+TIP: There's more about producers. You can declare qualifiers, inject dependencies into the producer methods parameters, etc. You can read more about producers for example in the https://docs.jboss.org/weld/reference/latest/en-US/html/beanscdi.html#_producer_methods[Weld docs, window="_blank"].
+
+Synthetic beans are usually provided by extensions.
+A synthetic bean is not backed by a class/method/field.
+Application developers usually only consume this kind of beans.
+
+== _Q: OK injection looks cool. What other services are provided?_
+
+=== Lifecycle Callbacks
+
+A bean class may declare lifecycle `@PostConstruct` and `@PreDestroy` callbacks:
+
+.Lifecycle Callbacks Example
+[source,java]
+----
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+
+@ApplicationScoped
+public class Translator {
+
+    @PostConstruct <1>
+    void init() {
+       // ...
+    }
+    
+    @PreDestroy <2>
+    void destroy() {
+      // ...
+    }
+}
+----
+<1> This callback is invoked before the bean instance is put into service. It is safe to perform some initialization here.
+<2> This callback is invoked before the bean instance is destroyed. It is safe to perform some cleanup tasks here.
+
+TIP: It's a good practice to keep the logic in the callbacks "without side effects", i.e. you should avoid calling other beans inside the callbacks.
+
```

Comment:
Maybe add a note that Quarkus support custom callback global to the application via https://quarkus.io/guides/lifecycle ?

Reply:
Hm, there is no relation between lifecycle callbacks and `StartupEvent/@Startup`.

Reply:
Right, one is bean lifecycle and the other application lifecycle. 
My suggestion was to add a note to distinguish between the two, but I'm OK to keep the doc as it is now.